### ISSUE TITLE
[Wayland] Cursor disappearing when displaying any application

### ISSUE THREAD

**Issue by Crystal4276** [qtile/qtile#3864]
### The issue:

qtile 0.22.1-1

The cursor is present at the start of qtile, i can move it around.  
It disappears as soon as i open/display any application

I use three monitors with 3 groups per monitor.
Config is here : https://github.com/Crystal4276/qtile
qtile.log is empty.

It's a bit light on information, i know. 
I don't know where to look for to get more hints about what's happening.

### Required:

- [X] I have searched past issues to see if this bug has already been reported.

**Comment by m-col** [comment]
Ah, annoying. Could you confirm that you have the latest pywlroots release? I suspect you do though. Is the cursor still functional when clicking and moving, but it's just rendering the image that is not happening? Does it change when you move it in or out of windows?

Would you be able to test https://github.com/qtile/qtile/pull/3862 to see if it fixes the issue?

**Comment by Crystal4276** [comment]
This is pywlroots version:
python-pywlroots 0.15.22-1
"the cursor still functional when clicking and moving, but it's just rendering the image that is not happening?"
Yes, exactly

"Does it change when you move it in or out of windows?"
Yes it does, i see windows borders activating when i move around.

"Would you be able to test https://github.com/qtile/qtile/pull/3862 to see if it fixes the issue?"
Let me try/have a look

**Comment by Crystal4276** [comment]
I apologize if what i will ask is absolutely stupid (i'm not a dev...)

For https://github.com/qtile/qtile/pull/3862 should i copy / paste brutally the content of the pull request to libqtile/backend/wayland/core.py on my system ?

I'm a bit afraid to do that.
There must be a better way to test it ?

**Comment by Crystal4276** [comment]
Ok that's probably not very elegant what i have done but at least i could test https://github.com/qtile/qtile/pull/3862

Results : 
- The cursor now does not appear at all even after qtile startup
- The cursor is block on screen 0 and cannot reach screen 1 and 2 on a triple monitor setup

**Comment by m-col** [comment]
> For https://github.com/qtile/qtile/pull/3862 should i copy / paste brutally the content of the pull request to libqtile/backend/wayland/core.py on my system ?

That would work, yes :) though see: http://docs.qtile.org/en/stable/manual/howto/git.html#i-want-to-try-out-a-feature-somebody-is-working-on. That might get you in the right direction for checking out a pull request in a locally cloned copy of the repo

Thanks for testing. Which distro are you using? Just to clarify - did this issue only appear now that you updated to 0.22.1?

**Comment by Crystal4276** [comment]
I'm on Arch.
This issue was already there with 0.21.
Just didn't took the time to report it, I was focused on widgets.

One question related to testing the latest pull requests:
 5 aur/qtile-wayland-git 0.20.0.r33.gc1f50739-2 [+1 ~0.28]
    A full-featured, pure-Python tiling window manager - Wayland. (git version)
 3 aur/qtile-git 0.20.0.r33.gc1f50739-2 [+70 ~0.22]
    A full-featured, pure-Python tiling window manager - X11. (git version)

If I install those two AUR packages and remove the stable qtile 0.22, i should be able to always get the latest version with all merged pull request ?
Anything in particular i should know before doing that ?

**Comment by m-col** [comment]
>  This issue was already there with 0.21.

Interesting. Do you have `xcb-util-cursor` installed? Are you using an Nvidia GPU?

> If I install those two AUR packages and remove the stable qtile 0.22, i should be able to always get the latest version with all merged pull requests ?
>  Anything in particular i should know before doing that ?

Yes, this will give you the latest commits. Note that this will be (even more) unstable compared to the latest release, but otherwise it should be fine. I think you'll only be able to install one of `qtile-git` or `qtile-wayland-git`, because they conflict (e.g. both install /usr/bin/qtile) but each has the dependencies of the other as optional deps IIRC, so just make sure you have all the deps you need.

**Comment by m-col** [comment]
If you run qtile with `WLR_NO_HARDWARE_CURSORS=1` in the environment, does the cursor appear?

**Comment by Crystal4276** [comment]
` extra/xcb-util-cursor 0.1.3-4 [0B 28.67KiB] [Installed]`
Nvidia:
```
  2 extra/egl-wayland 2:1.1.11-2 [0B 85.28KiB] [Installed]
    EGLStream-based Wayland external platform
  1 extra/nvidia 515.76-1 [0B 29.98MiB] [Installed]
    NVIDIA drivers for linux
```
Env variable set:
```
❯ printenv |grep -i wlr
WLR_NO_HARDWARE_CURSORS=1 
```
Cursor image still vanishing with opening applications. ( I used the original core.py)

**Comment by m-col** [comment]
I updated the PR - please could test with the updated version?

**Comment by Crystal4276** [comment]
Six files...  :(

I'll do it manually this time , but I guess it's time to learn git


OMessage ID: ***@***.***>

**Comment by m-col** [comment]
That is a shame. My current theory is that the issue is due to some incompatibility with your gpu or driver and wlroots, and if that is the case then there is nothing we can do to fix it. I suggest you try using the nouveau driver, and search https://github.com/swaywm/sway/issues for potential solutions/workarounds.

**Comment by Crystal4276** [comment]
I tried to install nouveau and removed nvidia driver but i cannot reach the display manager (black screen).
Maybe i'll try another DE that has wayland support to check if my hardware is supported (RTX2080).

My original motivation of looking into qtile-wayland was an attempt to avoid this issue with X11:
https://github.com/qtile/qtile/discussions/3839

Well, thanks you for taking the time to look into this.

**Comment by m-col** [comment]
>  Maybe i'll try another DE that has wayland support to check if my hardware is supported (RTX2080).

Great idea, can't believe I didn't think of that - if you can get sway (or any other wlroots compositor) working fine then we can rule out the GPU issue and confirm it is a qtile issue. That would be super helpful :)

**Comment by Crystal4276** [comment]
So i tried sway directly.
I had to use --unsupported-gpu option to start.
Results: no cursor image at all. (but still can activate different part of the screen like qtile).

I'm still curious to try gnome/kde wayland.

**Comment by ervinpopescu** [comment]
Do you have mesa installed? Most likely yes, as ["nvidia-utils depends on libglvnd which depends on mesa."](https://bbs.archlinux.org/viewtopic.php?pid=2008403#p2008403) so please post your version.

**Comment by Crystal4276** [comment]
That's all i have :
```
2 extra/glu 9.0.2-3 [0B 354.94KiB] [Installed]
    Mesa OpenGL utility library
 1 extra/mesa 22.1.7-1 [0B 72.32MiB] [Installed]
    An open-source implementation of the OpenGL specification
15 multilib/lib32-mesa 22.1.7-1 [0B 62.40MiB] [Installed]
    An open-source implementation of the OpenGL specification (32-bit)
```

**Comment by Crystal4276** [comment]
Are you saying you are facing the same bug with an AMD card ?
m-col was suggesting it is a 100% Nvidia issue.

**Comment by ervinpopescu** [comment]
Well it seems oddly similar to what I had before I downgraded to mesa 22.1.7, and that was all I did. My problems also involved:

- rendering issues(I don't think that's the right word) with nitrogen, rectangles of the image being moved to another part of the image, whatever that is

- cursor issues, sometimes it worked in electron apps, which is even weirder, it wouldn't work at all with qtile things such as the bar and popups, it sometimes got better with a qtile restart

- sddm issues:
  - random wallpaper, either black or the picture I chose
  - absent cursor 99% of the time
  - absent letters in the WM names

**Comment by Crystal4276** [comment]
@m-col (Hi!)
I'm still digging on the web for what i could do to solve this issue with nvidia GPU.

On Arch, I'd like to try wlroots-eglstreams instead of wlroots.
But Python-pywlroots requires wlroots as a dependency.
I'm struggling to find a way to point python-pywlroots towards wlroots-eglstreams using pacman.
Is that something even possible ?

**Comment by Crystal4276** [comment]
Nvidia has triaged a bug internally in July 2022:
https://github.com/NVIDIA/open-gpu-kernel-modules/issues/318
I guess we just have to wait.

**Comment by m-col** [comment]
> On Arch, I'd like to try wlroots-eglstreams in the AUR instead of wlroots.
> But python-pywlroots requires wlroots as a dependency.
> I'm struggling to find a way to point python-pywlroots towards wlroots-eglstreams using pacman.
> Is that something even possible ?

I'm not sure but in theory, if it's just a patched version of wlroots, then if it's the system's version of wlroots when pywlroots is built (e.g. when installed with pip) then it would use it.

>  I guess we just have to wait.

Yeah... it looks like discussion is active which is nice, so hopefully it will get resolved soon.

**Comment by Crystal4276** [comment]
Some config that could help to make it work.
From hyprland  which is also wlroots based.
https://wiki.hyprland.org/Nvidia/

It didn't for me though.

**Comment by Crystal4276** [comment]
Hello,

Just reporting back for this bug, because finally could upgrade to 0.23 on arch.
Since the first report, the nvidia driver are now at version 535.113.01, and i'm still on a RTX2080

Unfortunately, the bug is still there.

**Comment by yobleck** [comment]
Hello I finally managed to get Qtile to run under Wayland on my system but I have the same issue. My setup is very similar to @Crystal4276 but with Manjaro and an RTX 3070. All packages are the same version.
I've got basically the same error log
```
2023-10-14 14:38:32,865 ERROR wlroots log.py:log_func_callback():L15 [EGL] command: eglQueryDmaBufModifiersEXT, error: EGL_BAD_PARAMETER (0x300c), message: "EGL_BAD_PARAMETER error: In eglQueryDmaBufModifiersEXT: Invalid format"
2023-10-14 14:38:32,865 ERROR wlroots log.py:log_func_callback():L15 [render/egl.c:841] Failed to query dmabuf number of modifiers
2023-10-14 14:38:32,865 ERROR wlroots log.py:log_func_callback():L15 [EGL] command: eglQueryDmaBufModifiersEXT, error: EGL_BAD_PARAMETER (0x300c), message: "EGL_BAD_PARAMETER error: In eglQueryDmaBufModifiersEXT: Invalid format"
2023-10-14 14:38:32,865 ERROR wlroots log.py:log_func_callback():L15 [render/egl.c:841] Failed to query dmabuf number of modifiers
2023-10-14 14:38:33,460 WARNING libqtile prompt.py:_configure():L428 Prompt widget only supports audible bell under X11
2023-10-14 14:38:33,460 WARNING libqtile bar.py:_configure_widget():L358 Widget removed: Systray does not support <libqtile.backend.wayland.core.Core object at 0x7f834d9fbb10>.
2023-10-14 14:38:49,129 WARNING libqtile lifecycle.py:_atexit():L37 Qtile will now terminate
```

I launch Qtile from the terminal with `qtile start -b wayland -c /path/to/default/config` and open a terminal emulator like konsole or kitty. I have a dual monitor setup so the window opens on the other monitor and the cursor disappears as soon as I move it over the window. 

`WAYLAND_DEBUG=1` spits out tons of lines like 

```
wl_pointer@18.enter(9, wl_surface@25, 2552.51562500, 537.77734375)
wl_pointer@18.motion(3305951, 2144.55859375, 513.85546875)
wl_pointer@18.set_cursor(9, nil, 0, 0)
```
but I'n not sure if that means anything.

I should also note that riverwm also has the same issue with the cursor disappearing, so the issue is probably upstream with wlroots or the Nvidia driver.

**Comment by m-col** [comment]
@Crystal4276 that `AssertionError` is probably unrelated and can probably be ignored for now, though that shouldn't really happen. I think it means the core is getting set up earlier than qtile is ready for it. Is that happening only if you're moving the mouse while qtile starts up, then shortly after it doesn't happen?

> window.window.get_net_wm_pid()

You want `window.get_pid()`. `window.window.get_net_wm_pid()` is an x11 backend implementation detail, but `window.get_pid()` will work for any backend.

>  window.window.get_wm_class()

Similarly, you want just `window.get_wm_class()`.

re @yobleck:

> I should also note that riverwm also has the same issue with the cursor disappearing, so the issue is probably upstream with wlroots or the Nvidia driver.

Thanks for confirmation about riverwm. We don't do much in the way of this lower-level logic under the wlroots hood, so I also suspect this is a wlroots thing, or at least riverwm and us are doing something not quite right in the meantime.

**Comment by ErgoPraxis** [comment]
The `export WLR_NO_HARDWARE_CURSORS=1` placed in `/etc/profile` worked for me. 

I'm running a minimal install (no other DE on system) with just Qtile, Nvidia propriety driver, Arch Linux, with Wayland. 

The same behavior was observed: When initially starting the system and qtile load or loading, the cursor was movable and visible. Only have some random seconds would it disappear. Usually within 5 seconds. The cursor was functional just not visible; I was rather sherlock holmes like trying to close firefox with a correctly located click.

The above env var seems to work. I will post back here if I find otherwise. So far, cursor has been visible for 5 mins.

**Comment by Crystal4276** [comment]
Finally !
Issue solved with the installation of driver 555.58 stable on Arch.

Still some problem with explicit sync but hopefully wlroots 0.18 will solve it ! :-)

**Comment by Crystal4276** [comment]
Finally !
Issue solved with the installation of driver 555.58 stable on Arch.

Still some problem with explicit sync but hopefully wlroots 0.18 will solve it ! :-)
